dymos/phases/solve_ivp/components/solve_ivp_polynomial_control_group.py

```diff
@@ -0,0 +1,249 @@
+
+            self.declare_partials(of=self._output_rate_names[name],
+                                  wrt=self._input_names[name],
+                                  rows=self.rate_jac_rows[name], cols=self.rate_jac_cols[name])
+
+            self.declare_partials(of=self._output_rate2_names[name],
+                                  wrt='t_duration', rows=rs, cols=np.zeros_like(rs))
+
+            self.declare_partials(of=self._output_rate2_names[name],
+                                  wrt=self._input_names[name],
+                                  rows=self.rate2_jac_rows[name], cols=self.rate2_jac_cols[name])
+
+    def compute(self, inputs, outputs, discrete_inputs=None, discrete_outputs=None):
+
+        dt_dptau = 0.5 * inputs['t_duration']
+
+        for name, options in iteritems(self.options['polynomial_control_options']):
+            L_do, D_do, D2_do = self._matrices[name]
+
+            u = inputs[self._input_names[name]]
+
+            a = np.tensordot(D_do, u, axes=(1, 0)).T
+            b = np.tensordot(D2_do, u, axes=(1, 0)).T
+
+            # divide each "row" of the rates by dt_dptau or dt_dptau**2
+            outputs[self._output_val_names[name]] = np.tensordot(L_do, u, axes=(1, 0))
+            outputs[self._output_rate_names[name]] = (a / dt_dptau).T
+            outputs[self._output_rate2_names[name]] = (b / dt_dptau ** 2).T
+
+    def compute_partials(self, inputs, partials):
+        nn = self.options['grid_data'].num_nodes
+
+        for name, options in iteritems(self.options['polynomial_control_options']):
+            control_name = self._input_names[name]
+            num_input_nodes = options['order'] + 1
+            L_do, D_do, D2_do = self._matrices[name]
+
+            size = self.sizes[name]
+            rate_name = self._output_rate_names[name]
+            rate2_name = self._output_rate2_names[name]
+
+            # Unroll matrix-shaped controls into an array at each node
+            u_d = np.reshape(inputs[control_name], (num_input_nodes, size))
+
+            t_duration = inputs['t_duration']
+            t_duration_tile = np.tile(t_duration, size * nn)
+
+            partials[rate_name, 't_duration'] = \
+                0.5 * (-np.dot(D_do, u_d).ravel(order='F') / (0.5 * t_duration_tile) ** 2)
+
+            partials[rate2_name, 't_duration'] = \
+                -1.0 * (np.dot(D2_do, u_d).ravel(order='F') / (0.5 * t_duration_tile) ** 3)
+
+            t_duration_x_size = np.repeat(t_duration, size * nn)[:, np.newaxis]
+
+            r_nz, c_nz = self.rate_jac_rows[name], self.rate_jac_cols[name]
+            partials[rate_name, control_name] = \
+                (self.rate_jacs[name] / (0.5 * t_duration_x_size))[r_nz, c_nz]
+
+            r_nz, c_nz = self.rate2_jac_rows[name], self.rate2_jac_cols[name]
+            partials[rate2_name, control_name] = \
+                (self.rate2_jacs[name] / (0.5 * t_duration_x_size) ** 2)[r_nz, c_nz]
+
+
+class SolveIVPPolynomialControlGroup(Group):
+
```

Comment:
I guess my only quibble is the use of an action verb in all these class names.   It makes it sound like SolveIVPPolynomialControlGroup is a group that solves an IVP with polynomial controls, rather than this being a helper component for the SolveIVPPhase.  I don't have a suggestion that isnt more wordy though.

Reply:
One option would be removing the SolveIVP from its name, it would still live in the `phases.solve_ivp.components` namespace to separate it from the one in `phases.components`.  I generally think it's better to uniquely define its class name though.